lib/tasks/version_scheme_counter.rake

```diff
@@ -0,0 +1,108 @@
+# 3. If tied for unknown, mark unknown.
+# 4. Since calver validates as other schemes, prefer calver in the case of a tie since it is unusual and likely calver.
+# 5. In the case of a tie, if Semver is one of the contenders, prefer Semver for its ubiquity.
+# 6. If a tie between Maven and PEP440, check platform and choose accordingly.
+#
+# Known issues: many version schemes will match PEP440 and even Maven and yet be Semver. In the case of a tie, we choose
+#   semver, but this may turn out to be the wrong choice later on.
+
+module VersionSchemeDetection
+  REGEXES = {
+    MAVEN: /^(0|([1-9]\d*))(\.[1-9]\d*){1,2}(-(((alpha|beta|rc)-)?\d+|SNAPSHOT))?$/,
+    # PEP440 Regex copied from https://www.python.org/dev/peps/pep-0440/#appendix-b-parsing-version-strings-with-regular-expressions
+    PEP440: /^([1-9][0-9]*!)?(0|[1-9][0-9]*)(\.(0|[1-9][0-9]*))*((a|b|rc)(0|[1-9][0-9]*))?(\.post(0|[1-9][0-9]*))?(\.dev(0|[1-9][0-9]*))?$/
+  }
+
+  # validators return truthy/falsy
+  VALIDATORS = {
+    MAVEN: ->(version) { version =~ REGEXES[:MAVEN] },
+    PEP440: ->(version) { version =~ REGEXES[:PEP440] },
+    SEMVER: ->(version) { SemanticRange.valid(version).present? },
+    CALVER: lambda do |version|
+      first_part = version.split(".")[0]
+      # Assume most if not all Calver versions were released after 1000 AD
+      # Also assume if they are using YY as the first part that it is after 2000 AD
+      first_part.length >= 4 || (first_part.length == 2 && first_part.to_i > 0)
+    end
+  }
+end
+
+tallies = { semver: 0, pep440: 0, maven: 0, calver: 0, unknown: 0, no_versions: 0 }
+
+namespace :version do
+  desc 'Tests a sampling of project\'s versions to count occurrences of different versioning schemes'
+  task scheme_counter: :environment do
+    # projects = Project.order(Arel.sql("RANDOM()")).limit(100)
+
+    global_tallies = tallies.clone.merge({
+      unknown_versions: []
+    })
+
+    Project.find_each(batch_size: 1000) do |project|
+      local_tallies = tallies.clone
+
+      if !project.versions_count?
+        global_tallies[:no_versions] += 1
+        next
+      end
+
+      project.versions.each do |version|
+        version_number = version.number
+        matchable = false
+        [
+          :semver,
+          :pep440,
+          :maven,
+          :calver,
+        ].each do |scheme|
+          if VersionSchemeDetection::VALIDATORS[scheme.upcase].call(version_number)
+            local_tallies[scheme] += 1
+            matchable = true
+          end
+        end
+
+        if !matchable
+          local_tallies[:unknown] += 1
+        end
+      end
+
+      max_count = local_tallies.values.max
+      maxes = local_tallies.select{|_, tally| tally == max_count}.keys
+
+      if maxes.length == 1
+        detected_scheme = maxes[0]
+      elsif maxes.include?(:unknown)
+        global_tallies[:unknown_versions].push([project.platform, project.name, project.versions.pluck(:number)])
```

Comment:
If you do the `includes(:versions)` suggestion from above, could change the pluck to a map here too

Reply:
What is the difference between the two? I've seen `.map(&:number)` before but I thought it was only needed when referring to a non-getter method on the object?

Reply:
The `pluck()` does a new query for the column you need (`SELECT number FROM versions WHERE versions.package_id = ?`) and doesn't hydrate any ActiveRecord models, so it's just a list of values.

So if you don't do the `includes(:versions)` earlier, it won't have any preloaded versions and this'll be the most efficient way.

But if you `includes(:versions)` earlier, it'll preload your versions (and hydrate their ActiveRecord models) with all the numbers in a single query, and it'll already have the `:number` values in memory, so doing a `pluck(:number)` query for each project would be extraneous work.